packages/mds-compliance-service/repository/index.ts

```diff
@@ -146,12 +149,84 @@ class ComplianceSnapshotReadWriteRepository extends ReadWriteRepository {
     }
   }
 
+  public getComplianceViolationPeriods = async (
+    options: GetComplianceViolationPeriodsOptions
+  ): Promise<ComplianceViolationPeriodEntityModel[]> => {
+    const { start_time, end_time = now(), policy_ids = [], provider_ids = [] } = options
+    const { connect } = this
+    try {
+      const connection = await connect('ro')
+      const entityManager = getManager(connection.name)
+      const mainQueryPart1 = `select
+      provider_id, policy_id,
+      start_time,
+      end_time,
+      LEAD(end_time, 1, NULL) OVER (partition by provider_id, policy_id order by start_time) as real_end_time,
+      compliance_snapshot_ids,
+      sum_total_violations
+      from (
+        select
+          provider_id, policy_id,
+          min(compliance_as_of) as start_time,
+          max(compliance_as_of) as end_time,
+          array_agg(compliance_snapshot_id) as compliance_snapshot_ids,
+          sum(total_violations) as sum_total_violations
+          from (
+            select
+            provider_id, policy_id,
+            max(group_number) OVER (partition BY provider_id, policy_id order by compliance_as_of) as group_number,
+            compliance_as_of, compliance_snapshot_id, total_violations
+            from (
+                select
+                provider_id, policy_id, compliance_as_of, compliance_snapshot_id, total_violations,
+                CASE WHEN
+                LAG(total_violations) OVER (partition BY provider_id, policy_id order by compliance_as_of) > 0
+                  AND total_violations > 0
+                  THEN NULL
+                ELSE
+                  row_number() OVER (partition BY provider_id, policy_id order by compliance_as_of)
+                END group_number
+                from ${entityManager.getRepository(ComplianceSnapshotEntity).metadata.tableName}
+                where
+                  compliance_as_of >= $1
+                  and compliance_as_of <= $2
+                  `
+      const mainQueryPart2 = `
+                order by provider_id, policy_id, compliance_as_of
+            ) s1
+          ) s2
+          group by provider_id, policy_id, group_number
+        ) s3; `
+
+      const queryParams: (Timestamp | UUID[])[] = [start_time, end_time]
+      const queryPolicyIDs = policy_ids.length > 0
+      const queryProviderIDs = provider_ids.length > 0
+      const queryArray = [mainQueryPart1]
+
+      if (queryPolicyIDs) {
+        queryParams.push(policy_ids)
+        queryArray.push(`and policy_id = ANY($${queryParams.length})`)
+      }
+
+      if (queryProviderIDs) {
+        queryParams.push(provider_ids)
+        queryArray.push(`and provider_id = ANY($${queryParams.length})`)
+      }
+      queryArray.push(mainQueryPart2)
```

Comment:
Can this be abstracted to a function, and made less mutable? When I see tons of `push` calls, especially to separate arrays, and they are order dependent, I get suuuuper spooked 😬

Reply:
I'm not doing the tuple thing you suggested. It just doesn't work. You have to add the `policy_ids` and `provider_ids` to the same array that holds `start_time` and `end_time` to get the correct number for the positional parameter in the SQL query. There's no point in having something like `[[`and policy_id = ANY($3)`, policy_ids]]`.

Reply:
I don't think that tuple thing can work. There's no point in having a tuple that goes like `[[`and provider_id = ANY($3)], provider_ids]]` because you need to add `provider_ids` to an array containing `start_time` and `end_time` to know what number to use for the positional parameter in the SQL query. You could add 2 (1 for start_time and another for end_time) to the length of the tuple, but that's just stupid. 

I'm copying over the SqlVals class from `mds-db`. Hopefully it's less spooky now.